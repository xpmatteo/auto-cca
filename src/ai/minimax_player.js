import { score } from "./score.js";

class TreeNode {
    constructor(score) {
        this.score = score;
        this.children = [];
    }

    addChild(game, command) {
        const clone = game.clone();
        clone.executeCommand(command);
        const theScore = score(clone, clone.currentSide);
        this.children.push(new TreeNode(theScore));
    }
}

export class MinimaxPlayer {
    search(game, depth) {
        if (depth === 0) {
            const sco = score(game, game.currentSide);
            return new TreeNode(sco)
        }
        const rootNode = new TreeNode();
        game.validCommands().forEach((command) => {
            rootNode.addChild(game, command);
        });
        rootNode.score = -Infinity;
        for (let i = 0; i < rootNode.children.length; i++) {
            const child = rootNode.children[i];
            rootNode.score = Math.max(rootNode.score, child.score);
        }
        return rootNode;
    }
}
