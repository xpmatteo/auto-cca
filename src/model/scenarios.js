import { hexOf } from "../lib/hexlib.js";
import GameStatus from "./game_status.js";
import { Side } from "./side.js";
import * as units from "./units.js";

export class Scenario {
    /**
     * @param side
     * @returns {Side}
     */
    opposingSide(side) {
        return side === this.sideNorth ? this.sideSouth : this.sideNorth;
    }

    /**
     * @param {Game} game
     * @returns {GameStatus}
     */
    gameStatus(game) {
        if (game.killedUnitsOfSide(Side.CARTHAGINIAN).length === this.pointsToWin) {
            return GameStatus.ROMAN_WIN;
        }
        if (game.killedUnitsOfSide(Side.ROMAN).length === this.pointsToWin) {
            return GameStatus.CARTHAGINIAN_WIN;
        }
        return GameStatus.ONGOING;
    }

    /**
     * @param {Board|Game} board
     */
    placeUnitsOn(board) {
        throw new Error("Subclass must implement");
    }
}

export class NullScenario extends Scenario {
    firstSide = Side.ROMAN;
    sideNorth = Side.CARTHAGINIAN;
    sideSouth = Side.ROMAN;
    pointsToWin = 3;
    commandNorth = 1;
    commandSouth = 1;

    placeUnitsOn(board) {
    }

    gameStatus(board) {
        return GameStatus.ONGOING;
    }
}

export class MeleeScenario extends Scenario {
    firstSide = Side.ROMAN;
    sideNorth = Side.CARTHAGINIAN;
    sideSouth = Side.ROMAN;
    pointsToWin = 3;
    commandNorth = 4;
    commandSouth = 3;

    placeUnitsOn(board) {
        // 1522
        board.placeUnit(hexOf(0, 5), new units.RomanHeavyInfantry());
        board.placeUnit(hexOf(1, 5), new units.RomanHeavyInfantry());
        board.placeUnit(hexOf(2, 5), new units.RomanHeavyInfantry());
        board.placeUnit(hexOf(3, 5), new units.RomanHeavyInfantry());

        // 6422
        // board.placeUnit(hexOf(1, 4), new units.RomanHeavyInfantry());
        // board.placeUnit(hexOf(2, 4), new units.RomanHeavyInfantry());
        // board.placeUnit(hexOf(3, 4), new units.RomanHeavyInfantry());
        // board.placeUnit(hexOf(4, 4), new units.RomanHeavyInfantry());

        board.placeUnit(hexOf(2, 3), new units.CarthaginianHeavyInfantry());
        board.placeUnit(hexOf(4, 2), new units.CarthaginianHeavyInfantry());
        board.placeUnit(hexOf(4, 3), new units.CarthaginianHeavyInfantry());
    }
}

export class OneToOneMeleeScenario extends Scenario {
    firstSide = Side.ROMAN;
    sideNorth = Side.CARTHAGINIAN;
    sideSouth = Side.ROMAN;
    pointsToWin = 1;
    commandNorth = 3;
    commandSouth = 3;

    placeUnitsOn(board) {
        board.placeUnit(hexOf(2, 2), new units.CarthaginianHeavyInfantry());
        board.placeUnit(hexOf(1, 4), new units.RomanLightBowsInfantry());
    }
}

export class TwoOnTwoMeleeScenario extends Scenario {
    firstSide = Side.ROMAN;
    sideNorth = Side.CARTHAGINIAN;
    sideSouth = Side.ROMAN;
    pointsToWin = 2;
    commandNorth = 4;
    commandSouth = 3;

    placeUnitsOn(board) {
        board.placeUnit(hexOf(-1, 6), new units.RomanHeavyInfantry());
        board.placeUnit(hexOf(0, 6), new units.RomanHeavyInfantry());

        board.placeUnit(hexOf(0, 4), new units.CarthaginianHeavyInfantry());
        board.placeUnit(hexOf(1, 4), new units.CarthaginianHeavyInfantry());
    }
}


export class AkragasScenario extends Scenario {
    firstSide = Side.SYRACUSAN;
    sideNorth = Side.CARTHAGINIAN;
    sideSouth = Side.SYRACUSAN;
    pointsToWin = 5;
    commandNorth = 5;
    commandSouth = 6;

    placeUnitsOn(board) {
        board.placeUnit(hexOf(1, 1), new units.CarthaginianMediumInfantry());
        board.placeUnit(hexOf(2, 1), new units.CarthaginianAuxiliaInfantry());
        board.placeUnit(hexOf(3, 1), new units.CarthaginianHeavyInfantry());
        board.placeUnit(hexOf(4, 1), new units.CarthaginianMediumInfantry());
        board.placeUnit(hexOf(6, 1), new units.CarthaginianAuxiliaInfantry());
        board.placeUnit(hexOf(7, 1), new units.CarthaginianAuxiliaInfantry());
        board.placeUnit(hexOf(8, 1), new units.CarthaginianAuxiliaInfantry());

        board.placeUnit(hexOf(1, 2), new units.CarthaginianHeavyChariot());
        board.placeUnit(hexOf(3, 2), new units.CarthaginianLightBowsInfantry());
        board.placeUnit(hexOf(7, 2), new units.CarthaginianLightInfantry());
        board.placeUnit(hexOf(9, 2), new units.CarthaginianHeavyChariot());
        board.placeUnit(hexOf(10, 1), new units.CarthaginianLightCavalry());

        function placeSyracusanUnit(hex, unit) {
            board.placeUnit(hex, unit);
            unit.side = Side.SYRACUSAN;
        }
        placeSyracusanUnit(hexOf(-2, 7), new units.RomanLightCavalry());
        placeSyracusanUnit(hexOf(-1, 7), new units.RomanAuxiliaInfantry());
        placeSyracusanUnit(hexOf(0, 7), new units.RomanHeavyInfantry());
        placeSyracusanUnit(hexOf(2, 7), new units.RomanHeavyInfantry());
        placeSyracusanUnit(hexOf(3, 7), new units.RomanHeavyInfantry());
        placeSyracusanUnit(hexOf(4, 7), new units.RomanHeavyInfantry());
        placeSyracusanUnit(hexOf(5, 7), new units.RomanAuxiliaInfantry());
        placeSyracusanUnit(hexOf(6, 7), new units.RomanAuxiliaInfantry());
        placeSyracusanUnit(hexOf(7, 7), new units.RomanMediumCavalry());

        placeSyracusanUnit(hexOf(0, 6), new units.RomanLightBowsInfantry());
        placeSyracusanUnit(hexOf(3, 6), new units.RomanLightBowsInfantry());
        placeSyracusanUnit(hexOf(7, 6), new units.RomanLightInfantry());
    }
}

const SCENARIOS = {
    akragas: new AkragasScenario(),
    melee: new MeleeScenario(),
    oneToOneMelee: new OneToOneMeleeScenario(),
    twoOnTwoMelee: new TwoOnTwoMeleeScenario(),
}

export function makeScenario(name) {
    return SCENARIOS[name] || new AkragasScenario();
}
