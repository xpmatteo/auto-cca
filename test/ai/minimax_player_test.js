import makeGame from "model/game.js";
import { AkragasScenario, MeleeScenario } from "model/scenarios.js";
import { MinimaxPlayer } from "ai/minimax_player.js";


test('level zero search', () => {
    const game = makeGame(new AkragasScenario());
    const player = new MinimaxPlayer(0);
    const tree = player.search(game, 0);

    // expect(tree.score).toBe(9);
    expect(tree.children.length).toBe(0);
    expect(tree.bestCommands()).toEqual([]);
    expect(tree.size()).toBe(1);
});

test('level 1 search', () => {
    const game = makeGame(new AkragasScenario());
    const player = new MinimaxPlayer(game);
    const tree = player.search(game, 1);

    // expect(tree.score).toBe(9);
    expect(tree.children.length).toBe(3);
});

test('bestcommands', () => {
    const game = makeGame(new MeleeScenario());
    const player = new MinimaxPlayer(3);
    const best = player.decideMove(game);

    expect(best.length).toBe(3);
});


// test: stop when action is not deterministic
// test: stop when acion changes current side

// level 1: playCard
// level 2: activate units
// level 3: move units
test('search', () => {
    const game = makeGame(new AkragasScenario());
    const player = new MinimaxPlayer(game);
    const tree = player.search(game, 3);

    expect(tree.toString()).toMatchSnapshot();
});
