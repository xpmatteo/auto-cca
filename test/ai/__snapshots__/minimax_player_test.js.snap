// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`search 1`] = `"[object Object]"`;
